best_practices.rst

```diff
@@ -0,0 +1,474 @@
+of email notifications, or the enabled `feature toggles`_. Their value doesn't
+change per machine, so it's not worth it to define them as environment variables.
+
+Define these options as :ref:`parameters <configuration-parameters>` in the
+``config/services.yaml`` file. You can override these options per
+:ref:`environment <configuration-environments>` in the ``config/services_dev.yaml``
+and ``config/services_prod.yaml`` files.
+
+Use Short and Prefixed Parameter Names
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Consider using ``app.`` as the prefix of your :ref:`parameters <configuration-parameters>`
+to avoid collisions with Symfony and third-party bundles/libraries parameters.
+Then, use just one or two words to describe the purpose of the parameter:
+
+.. code-block:: yaml
+
+    # config/services.yaml
+    parameters:
+        # don't do this: 'dir' is too generic and it doesn't convey any meaning
+        app.dir: '...'
+        # do this: short but easy to understand names
+        app.contents_dir: '...'
+        # it's OK to use dots, underscores, dashes or nothing, but always
+        # be consistent and use the same format for all the parameters
+        app.dir.contents: '...'
+        app.contents-dir: '...'
+
+Use Constants to Define Options that Rarely Change
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Configuration options like the number of items to display in some listing rarely
+change. Instead of defining them as :ref:`service container parameters <configuration-parameters>`,
+define them as PHP constants in the related classes. Example::
+
+    // src/Entity/Post.php
+    namespace App\Entity;
+
+    class Post
+    {
+        public const NUMBER_OF_ITEMS = 10;
+
+        // ...
+    }
+
+The main advantage of constants is that you can use them everywhere, including
+Twig templates and Doctrine entities, whereas parameters are only available
+from places with access to the :doc:`service container </service_container>`.
+
+The only notable disadvantage of using constants for this kind of configuration
+values is that it's complicated to redefine their values in your tests.
+
+Business Logic
+--------------
+
+Don't Create any Bundle to Organize your Application Logic
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When Symfony 2.0 was released, Symfony applications used bundles to divide their
+code into logical features: UserBundle, ProductBundle, InvoiceBundle, etc.
+
+However, a bundle is meant to be something that can be reused as a stand-alone
+piece of software. If UserBundle cannot be used "as is" in other Symfony
+applications, then it shouldn't be its own bundle. Moreover, if InvoiceBundle
+depends on ProductBundle, then there's no advantage to having two separate bundles.
+
+Symfony applications can still use third-party bundles (installed in ``vendor/``)
+to add features, but you should use PHP namespaces instead of bundles to organize
+your own code.
```

Comment:
Perhaps we should make it clear that it's still ok to create internal/private bundles that are shared among different projects. I get that question quite often in conferences. People really think that creating bundles is not possible anymore.